# متغیرها و کار با آنها

تا الان با متغیرهای اعشاری، عدد صحیح و استرینگ آشنا شدیم. متغیرهای اعشاری و عدد صحیح شامل متغیرهای عددی می شوند. یکی دیگر از متغیرهای عددی اعداد مختلط هستند.

In [16]:
z = 3 + 4j
print (type(z))
real_part = z.real
imag_part = z.imag
print(real_part)
print(imag_part)

<class 'complex'>
3.0
4.0


در کد بالا زد (زی) عدد مختلط است که در آن 3 عدد حقیقی و 4جی عدد موهومی است. در پایتون از جی برای نوشتن قسمت موهومی استفاده می کنیم نه از آی. روش دیگری نیز برای وارد کردن عدد مختلط به عنوان داده وجود دارد. 

In [22]:
z1 = complex(5, 6)
print(type(z1))
real = z1.real
imag = z1.imag
print("بخش حقیقی = ", real, "بخش موهومی = ", imag)

<class 'complex'>
بخش حقیقی =  5.0 بخش موهومی =  6.0


پس اعداد مختلط نیز از انواع متغیرها هستند. در جلسه گذشته با بولین ها هم آشنا شدیم منتهی بولین ها فقط به رابطه == ختم نمی شوند بلکه روابط مختلف دیگری وجود دارد. 

In [30]:
a = 5 #گزاره اولیه
print (1, a != 5) #برابر نیست
print(2, a < 6) #کوچکتر بودن
print(3, a <= 3) #کوچتکر مساوی
print(4, a > 6) #بزرگتر بودن
print(5, a >= 5) #بزرگتر مساوی
print(6, a == 10) #مساوی بودن

1 False
2 True
3 False
4 False
5 True
6 False


متغیری دیگر وجود دارد که در چارچوب بولین‌ها در حالت == خودکار نادرست شناخته می‌شود و در حالت =! خودکار درست و به آن متغیر از نوع هیچ می‌گوییم

In [37]:
b = None 
print(b)
print(type(b))
print("A", b == 5)
print("B", b != 9)

None
<class 'NoneType'>
A False
B True


همانطور که ملاحظه می کنید متغیر بی برابر با 5 نیست حکمی اشتباه است و اینکه بگوییم برابر با 9 نیست حکمی درست است. نکته ای که وجود دارد این است که بزرگتر و کوچکتر بودن و بزرگتر مساوی و کوچکتر مساوی برای مقایسه با سایر اعداد ممکن نیست. اما چرا؟ 

In [40]:
print("C", b < 8)
print("D", b > 4)
print("E", b <= 7)
print("F", b >= 2)

TypeError: '<' not supported between instances of 'NoneType' and 'int'

هیچ بودن مقدار یک متغیر یعنی آن متغیر وجود ندارد و برای نمونه اگر هویت عدد 3.4 اعشاری باشد، هویت هیچ، هیج است و قابل مقایسه با سایر متغیرها نیست. و در واقع زمانی که متغیری هیچ است، ثابتی خاص است که نماینده نبودن مقداری است. پس علت اینکه پایتون نمی تواند سایر عملیات ها را به جز برابری و نابرابری در مقابل هیچ اجرا کند نبود تشابه در نوع متغیرها است. همچنین دقت داشته باشید که بولین‌ها برای درک مقادیر متغیرها بکار می‌روند نه هویت آنها و وقتی از == بین یک متغیر هیچ و عدد صحیح استفاده می‌کنیم، پایتون نگاه مقادیر آنان می‌کند نه هویت آنها و به همین دلیل == و =! را اجرا می‌کند.
اما سوالی که پیش می‌آید این است: آیا در کدهای زیر درستی یا نادرستی واقعا معنی دار است؟ 

In [146]:
b = None 
print(b)
print(type(b))
print("1", b == 5)
print("2", b != 9)

None
<class 'NoneType'>
1 False
2 True


در واقع در کد 1  از پایتون می خواهیم که مقدار متغیر بی را بررسی کند و ببیند که برابر با 5 (که عدد صحیح است) هست یا نه. اما هویت متغیر بی هیچ است و نادرست بودن کد 1 با اینکه به ظاهر بی‌ایراد است کاملا دقیق نیست. در کد 2 این مسئله شفاف‌تر می‌شود. کد 2 می‌گوید که بی برابر با 9 نیست که صادق است ولی وقتی برابر 9 نیست خود بی چیست؟ آیا نمی‌تواند 8 باشد؟ یا متغیر استرینگ باشد؟ 8 نیز با 9 برابر نیست و متغیر استرینگ نیز با 9 برابر نیست پس در این حالت انگار بی علاوه بر اینکه از نوع متغیر هیچ است، از نوع سایر متغیرها هم هست که اینکار دقیق نیست. راه حل چیست؟

In [43]:
team = None
team1 = 5
team2 = None

print(1, team is team1)
print(2, team is not team1)
print(3, team is team2)
print(4, team is not team2)

1 False
2 True
3 True
4 False


روش بالا مخصوص متغیرهای هیچ است و اینگونه دیگر به عنوان متغیر بولین شناخته نمی‌شوند و مقدار آنها در نظر گرفته نمی‌شود و هویتشان نیز در کدها حفظ می‌شود. این روش سه برتری کلی نسبت به روش قبل دارد:

 شفافیت: با استفاده از ایز و ایز نات دقیقا واضح است که داریم هویت متغیر را بررسی می‌کنیم نه مقدارش را.
 
 اطمینان داشتن: وقتی از == یا =! استفاده کنیم پایتون مقادیر را بررسی می‌کند و هویت متغیر را نادیده می گیرد و با اینکه جوابی درست می‌دهد ولی جوابش دقیق نیست.
 
 عملکرد: از نظر برنامه نویسی دستورهای بولین کندتر از دستورهای هویتی ایز و ایز نات اجرا می‌شوند

# (TypeCasting) تایپ کستینگ 

In [51]:
name = "Sara"
family_name = "Anderson"
age = 5

print(1, name + family_name)
print(2, name + family_name + age)

1 SaraAnderson


TypeError: can only concatenate str (not "int") to str

In [145]:
print(3,name+ " " + family_name + "is" + age +"years old")

TypeError: can only concatenate str (not "int") to str

همانطور که در جلسه پیش آمد، دو نوع متغیر متفاوت را نمی‌توان با هم برای مثال جمع کرد. برای اینکار باید نوع متغیرها را یکسان کنیم و سپس پرینت بگیریم. به این یکسان سازی تایپ کستینگ گفته می‌شود. در کدهای 2 و 3 متغیر اِیج عدد صحیح است اما سایر متغیرها استرینگ هستند.

In [54]:
print(4, name + family_name + str(age))
print(5,name+ " " + family_name + " is " + str(age) +" years old")

4 SaraAnderson5
5 Sara Anderson is 5 years old


در کد 4 و 5 متغیر اِیج را به استرینگ تبدیل کردیم و سپس پرینت گرفتیم و همانطور که مشاهده می‌شود خطا برطرف شد. 
راه دیگر استفاده از ، بود

In [55]:
print(name, family_name, "is", age, "years old")

Sara Anderson is 5 years old


اما مشکل اصلی در زمانی است که از کاربر داده می‌گیریم

In [85]:
Class_Number = input()
Class_Name = input()
Pi = input()
complex_number = input("Enter a complex number (e.g., 3+4j): ")

 103
 Python for Macroeconomics
 3.14
Enter a complex number (e.g., 3+4j):  4+8j


In [87]:
print("الف", type(Class_Number))
print("ب", type(Class_Name))
print("پ", type(Pi))
print("ت", type(complex_number))

الف <class 'str'>
ب <class 'str'>
پ <class 'str'>
ت <class 'str'>


همانطور که مشاهده می‌شود هر نوع داده‌ای، به عنوان استرینگ شناخته می‌شود. و اصولا می‌توان همه آنها را کنار هم به وسیله + پرینت گرفت ولی نمی‌توان عددی را با عددی از کاربر گرفتیم جمع کنیم، چون هر داده ای که از کاربر بگیریم استرینگ ("") خواهد بود.

In [86]:
print(Class_Number+Class_Name+Pi+complex_number)

103Python for Macroeconomics3.144+8j


In [84]:
summ = Class_Number+5
print(summ)

TypeError: can only concatenate str (not "int") to str

 برای تغییر هویت آنها باید از روشی که در قسمت قبل استفاده کردیم استفاده کنیم

In [89]:
print(float(Pi)+5)
print(complex(complex_number)+7+8j)

8.14
(11+16j)


اما دو نکته وجود دارد اول اینکه می‌توان مستقیما داده‌ای که از کاربر می‌گیریم را تغییر هویت دهیم.

In [71]:
your_age = int(input("how old are you = "))
print(type(your_age))

how old are you =  21


<class 'int'>


و نکته دوم در خصوص اعشاری کردن عدد پای است. در آن مثال درست است که 3.14 شبیه اعداد اعشاری است ولی در واقع جز اعداد متعالی است و باید به روشی دیگر آن را در پایتون نوشت 

In [82]:
import math

print(math.pi)
diameter = 10
radius = diameter/2
area_of_circle = (math.pi) * radius**2
circumference_of_circle = 2 * (math.pi) * radius

print("قطر= ", diameter)
print("شعاع= ", radius)
print("مساحت دایره= ", area_of_circle)
print("محیط دایره= ", circumference_of_circle)

3.141592653589793
قطر=  10
شعاع=  5.0
مساحت دایره=  78.53981633974483
محیط دایره=  31.41592653589793


در کد بالا برای وارد کردن عدد پای از کتابخانه‌های پایتون استفاده کردیم و آن را فراخوانی کردیم (ایمپورت کردن) و سپس محاسبات را انجام دادیم. پس برای اعداد متعالی نمی‌توانیم از هویت اعشاری استفاده کنیم.

# توابع

In [ ]:
print()
str()
float()
int()
input()
type()

توابع دستوراتی هستند که برای پایتون تعریف شده یا تعریف می‌شوند تا پایتون آنها را اجرا کند. به چیزهایی که یک تابع دریافت می‌کند تا عملیات تعریف شده‌ای را انجام دهد، آرگومان می‌گویند. تابع‌های بالا تابع‌‌های خود پایتون هستند در آدرس زیر تمامی تابع‌های خود پایتون را می‌توانید بیابید.

https://docs.python.org/3/library/functions.html

In [92]:
e1 = 3.9
e2 = 3

print(int(e1))
print(float(e2))

3
3.0


در نمونه‌های بالا برای تابع عدد صحیح تعریف شده که اعدادی که می‌گیرد به سمت پایین‌ترین عدد صحیح روند کند. می‌توان از تابع هلپ و یا داکسترینگ‌ جهت آگاه شدن از آرگومان‌هایی که یک تابع دریافت می‌کند بهره برد.  

In [ ]:
help(int) #چون نتیجه طولانی است آن را اجرا نمی‌کنم، در صورت تمایل توابع مختلف را خودتان به این شکل بررسی کنید

In [98]:
print(int.__doc__)

int([x]) -> integer
int(x, base=10) -> integer

Convert a number or string to an integer, or return 0 if no arguments
are given.  If x is a number, return x.__int__().  For floating point
numbers, this truncates towards zero.

If x is not a number or if base is given, then x must be a string,
bytes, or bytearray instance representing an integer literal in the
given base.  The literal can be preceded by '+' or '-' and be surrounded
by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
Base 0 means to interpret the base from the string as an integer literal.
>>> int('0b100', base=0)
4


در داکسترینگ بالا مشاهده می‌شود که تابع عدد صحیح دو آرگومان دریافت می‌کند، عدد و پایه و همینطور که در ادامه نوشته این تابع استرینگ نیز دریافت می‌کند و پایه تابع به صورت عادی 10 است (از صفر تا ده، ده عدد وجود دارد) و حداکثر پایه‌ای که می‌توان داد 36 است (10 به علاوه تعداد حروف الفبای انگلیسی).

In [123]:
e3 = "ab"

print(int(e3, 12))
print(int(e3, 36))
print(int(e3))

131
371


ValueError: invalid literal for int() with base 10: 'ab'

در کدهای بالا برای متغیر ایی3 یکبار پایه 12 در نظر گرفته و کد اجرا شده و عدد متناظر با آن که 131 پرینت گشته و در کد همان متغیر با پایه‌ای متفاوت به عدد صحیح تغییر یافته که عددی متفاوت با قبلی است. کد سوم به دلیل درست نبودن پایه 10 اجرا نشده. همانطور که گفته شد  پایه به صورت عادی 10 است. پس برای اجرا شدن کد اول باید پایه‌ای داده شود تا ای و بی شناخته شود که حداقل پایه‌ای که می‌توان داد 12 است.  

## چرا از توابع استفاده کنیم؟

توابع ما را از نوشتن مجدد چندین خط کد بی‌ نیاز می‌کنند و تنها با نوشتن یک دستور هدفی که می‌خواهیم را برآورده می‌کنند. در برنامه‌هایی که می‌نویسیم شاید نیاز پیدا کنیم که کاری مکررا انجام شود که در این صورت باید از چگونگی ساخت و اجرای تابع‌ها آگاه باشیم. 

## چگونه خودمان تابع تعریف کنیم؟

In [150]:
def Hello(): #def stands for define
    print("Hi") #Indentation
Hello()

Hi


تعریف تابع مطابق کد بالا انجام می‌شود. در پرانتز اول می‌توانیم آرگومان‌های تابع را مشخص کنیم. وقتی آرگومانی ندارد دستوری که به آن داده‌ایم پرینت می‌شود. بعد از تعریف اولیه نوشتن : الزامی است و فاصله‌ای که در خط بعدی ایجاد می‌شود چهار اسپیس یا یک تب است و به دلایل زیر ایجادش می‌کنیم:

- مشخص شدن بلاک تابع
- جدا کردن بلاک‌های تو در تو (بلاک‌های تو در تو در آینده توضیح داده می‌شود)
- حفظ شدن چارچوب: برای نمونه، متغیری که در بلاک تابع مشخص می‌شود با متغیری که خارج از آن تعریف می‌شود متفاوت است و در هنگام خواندن کد این چارچوب‌ها مهم است
- بهتر خوانده شدن کد

In [134]:
def Hello(name):
    print("Hello", name)
Hello("David")

Hello David


برای تابع بالا یک آرگومان تعریف شده و بعد از وارد شدن آن دستور در کنار آرگومان وارد شده می‌آید 

In [158]:
def adddition(a, b):
    result = a + b
    return result

In [162]:
adddition(5,4)

9

در تابع بالا دستور مربوط به جمع دو متغیر نوشته شده که در آن متغیری دیگر به نام ریزالت ساخته شده که از حاصل جمع دو آرگومان تابع است. دستور ریترن در پایان، متغیر ساخته شده را باز می‌گرداند. و در نهایت با وارد کردن آرگومان‌های تابع، دستور تابع اجرا می‌شود. 

## تابع تولید کاب داگلاس

In [177]:
# f(L,K) = A(L^a)*(K^b) 0<a<1 , 0<b<1

In [173]:
def cobb_douglas(L, K):
    ''' f(L,K) = A(L^a)*(K^b) 0<a<1 , 0<b<1'''
    A = 2
    a = 0.5
    b = 0.5
    value_of_function = A * (L**a) * (K**b)
    return value_of_function

L = float(input("how many labours do you have? = "))
K = float(input("how much capital do you have? = "))

output = cobb_douglas(L, K)
print("The value of the Cobb-Douglas function is:", output)
print(cobb_douglas.__doc__)

how many labours do you have? =  12
how much capital do you have? =  50


The value of the Cobb-Douglas function is: 48.98979485566356
 f(L,K) = A(L^a)*(K^b) 0<a<1 , 0<b<1


تابع کاب داگلاس فوق به گونه‌ای تعریف شده که متغیرهای سرمایه و نیروی کار از کاربر دریافت شود و چون می‌دانیم تابع اینپوت کلا استرینگ می‌دهد از تابع اعشاری استفاده شده تا داده دریافتی به عنوان عدد در نظر گرفته شود و در اجرای عملیات ریاضی اختلال ایجاد نکند. 

# Control Flow Structures (ساختارهای کنترل جریان) Part 1

### Conditional Statements (گزاره‌های شرطی)

ساختارهای کنترل جریان، ساختارهایی‌اند که بر ترتیب اجرا شدن کدها اثر می‌گذارند. این کار از طریق بررسی شرط‌ها و حلقه‌ها و تابع‌ها انجام می‌شود. برای مثال کد زیر را در نظر بگیرید

In [181]:
word = input("Please enter a four-letter word: ")
word_length = len(word)
if word_length == 4:
    print(word, "is a four-letter word. Well done")
else: print(word, "is not a four-letter word.")

Please enter a four-letter word:  hello


hello is not a four-letter word.


در کد بالا تعداد حروف کلمه‌ای که از کاربر گرفته‌ایم بررسی می‌شود و اگر تعداد حروف با 4 باشد عبارتی به خصوص پرینت می‌شود و اگر نباشد عبارتی دیگر.

- اگر = if
- اگر نباشد = else

تابعی جدیدی که در این کد ارائه شده تابع 
- len

است که کاربرد آن را گفتیم.

In [185]:
word1 = input("Please enter a four-letter word: ")
word_length1 = len(word1)
if word_length1 == 4:
    print(word1, "is a four-letter word. Well done")
elif word_length1 == 3:
    print(word1, "is a three-letter word. Try again")
else: print(word1, "is not a four-letter word.")

Please enter a four-letter word:  ask


ask is a three-letter word. Try again


در کد بالا شرط سومی در بین شرط اگر و اگر نباشد اضافه کردیم که از ترکیب شرط اول و سوم ایجاد می‌شود و امکان ایجاد شرط دیگری را نیز فراهم می‌آورد. 

In [1]:
x = 1
while x < 10:
    print(x)
    x = x + 1

1
2
3
4
5
6
7
8
9


روش دیگر برای ساختار بندی کد استفاده از دستور وایل است که ویژگی آن این است که تا وقتی که نتیجه حاصل شود به پردازش ادامه می‌دهد. برای مثال اگر 

- x = x+1

را در کد بالا حذف کنیم تا بی نهایت 1 پرینت خواهد شد چون کوچکتر از 10 است. 

# تمرین

- محیط و مساحت دایره‌ای را با گرفتن شعاع از کاربر حساب کنید

- ارزش آتی مبلغ 100 دلار را با نرخ بهره مرکب 5 درصد برای 6 سال را در دو حالت سالانه و شش ماهه محاسبه کنید.

برای حالت شش ماهه از فرمول زیر استفاده کنید
 
 S = Pe^rt


- سن دو نفر را بگیرید. اگر همسن بودند عبارت زیر پرینت شود:

 You are the same age

اگر همسن نبودند عبارت زیر پرینت شود

 You are not the same age

- تابعی تعریف کنید که زوج بودن یا فرد بودن یک عدد را بررسی کند.
اگر زوج بگوید زوج است و اگر فرد بود بگوید فرد است.